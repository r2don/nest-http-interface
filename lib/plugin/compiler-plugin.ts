import type * as ts from 'typescript';
import { mergePluginOptions } from './merge-options';
import { HttpInterfaceVisitor } from './visitors/http-interface.visitor';

const httpInterfaceVisitor = new HttpInterfaceVisitor();

function isFilenameMatched(patterns: string[], filename: string): boolean {
  return patterns.some((path) => filename.includes(path));
}

export const before: (
  options: Record<string, any> | undefined,
  program: ts.Program,
) => (ctx: ts.TransformationContext) => ts.Transformer<any> = (
  options,
  program,
) => {
  const mergedOption = mergePluginOptions(options);

  return (ctx: ts.TransformationContext): ts.Transformer<any> => {
    return (sf: ts.SourceFile) => {
      if (
        isFilenameMatched(
          mergedOption.interfaceFilenameSuffix as string[],
          sf.fileName,
        )
      ) {
        return httpInterfaceVisitor.visit(sf, ctx, program);
      }

      return sf;
    };
  };
};
