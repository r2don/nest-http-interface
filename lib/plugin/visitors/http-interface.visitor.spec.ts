import * as ts from 'typescript';
import { describe, expect, test } from 'vitest';
import {
  returnStringServiceCode,
  hasResponseBodyServiceCode,
  needResponseBodyServiceCode,
  notPromiseServiceCode,
  arrayResponseBodyServiceCode,
} from '../../fixtures/fake-service-code';
import { before } from '../compiler-plugin';

describe('HttpInterfaceVisitor', () => {
  const compilerOptions = {
    module: ts.ModuleKind.CommonJS,
    target: ts.ScriptTarget.ES2020,
    newLine: ts.NewLineKind.LineFeed,
    noEmitHelpers: true,
    experimentalDecorators: true,
    strict: true,
  };

  test('should ignore if file name is not match', () => {
    // given
    const filename = 'not-match.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);

    // when
    const result = ts.transpileModule(needResponseBodyServiceCode, {
      compilerOptions,
      fileName: filename,
      transformers: { before: [before({}, fakeProgram)] },
    });

    // then
    expect(result.outputText).toMatchSnapshot();
  });

  test('should ignore if return type is not a class', () => {
    // given
    const filename = 'text.service.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);

    // when
    const result = ts.transpileModule(returnStringServiceCode, {
      compilerOptions,
      fileName: filename,
      transformers: {
        before: [before(undefined, fakeProgram)],
      },
    });

    // then
    expect(result.outputText).toMatchSnapshot();
  });

  test('should ignore if method has ResponseBody decorator', () => {
    // given
    const filename = 'text.service.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);

    // when
    const result = ts.transpileModule(hasResponseBodyServiceCode, {
      compilerOptions,
      fileName: filename,
      transformers: { before: [before({}, fakeProgram)] },
    });

    // then
    expect(result.outputText).toMatchSnapshot();
  });

  test('should ignore if return type if not a promise ', () => {
    // given
    const filename = 'text.service.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);

    // when
    const result = ts.transpileModule(notPromiseServiceCode, {
      compilerOptions,
      fileName: filename,
      transformers: { before: [before({}, fakeProgram)] },
    });

    // then
    expect(result.outputText).toMatchSnapshot();
  });

  test('should override plugin suffix option', () => {
    // given
    const filename = '.custom.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);

    // when
    const result = ts.transpileModule(needResponseBodyServiceCode, {
      compilerOptions,
      fileName: filename,
      transformers: {
        before: [
          before({ interfaceFilenameSuffix: '.custom.ts' }, fakeProgram),
        ],
      },
    });

    // then
    expect(result.outputText).toMatchSnapshot();
  });

  test('should handle array return type', () => {
    // given
    const filename = 'text.service.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);

    // when
    const result = ts.transpileModule(arrayResponseBodyServiceCode, {
      compilerOptions,
      fileName: filename,
      transformers: {
        before: [before({}, fakeProgram)],
      },
    });

    // then
    expect(result.outputText).toMatchSnapshot();
  });
});
