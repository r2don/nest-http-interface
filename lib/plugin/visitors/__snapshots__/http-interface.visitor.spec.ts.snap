// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HttpInterfaceVisitor > should handle array return type 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const r2don_http_module_1 = require(\\"@r2don/nest-http-interface\\");
const nest_http_interface_1 = require(\\"@r2don/nest-http-interface\\");
class ResponseClass {
}
let UserService = class UserService {
    async getUsers() {
        throw new Error('not implemented');
    }
    async getUserList() {
        throw new Error('not implemented');
    }
    async getUsersReadonly() {
        throw new Error('not implemented');
    }
};
__decorate([
    (0, nest_http_interface_1.GetExchange)(),
    r2don_http_module_1.ResponseBody(ResponseClass)
], UserService.prototype, \\"getUsers\\", null);
__decorate([
    (0, nest_http_interface_1.GetExchange)(),
    r2don_http_module_1.ResponseBody(ResponseClass)
], UserService.prototype, \\"getUserList\\", null);
__decorate([
    (0, nest_http_interface_1.GetExchange)(),
    r2don_http_module_1.ResponseBody(ResponseClass)
], UserService.prototype, \\"getUsersReadonly\\", null);
UserService = __decorate([
    (0, nest_http_interface_1.HttpInterface)()
], UserService);
"
`;

exports[`HttpInterfaceVisitor > should ignore if file name is not match 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const nest_http_interface_1 = require(\\"@r2don/nest-http-interface\\");
class ResponseClass {
}
let UserService = class UserService {
    async getUser() {
        throw new Error('not implemented');
    }
};
__decorate([
    (0, nest_http_interface_1.GraphQLExchange)()
], UserService.prototype, \\"getUser\\", null);
UserService = __decorate([
    (0, nest_http_interface_1.HttpInterface)()
], UserService);
"
`;

exports[`HttpInterfaceVisitor > should ignore if method has ResponseBody decorator 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const nest_http_interface_1 = require(\\"@r2don/nest-http-interface\\");
const user_entity_1 = require(\\"./user.entity\\");
let UserService = class UserService {
    async getUser() {
        throw new Error('not implemented');
    }
};
__decorate([
    (0, nest_http_interface_1.PostExchange)(),
    (0, nest_http_interface_1.ResponseBody)(user_entity_1.User)
], UserService.prototype, \\"getUser\\", null);
UserService = __decorate([
    (0, nest_http_interface_1.HttpInterface)()
], UserService);
"
`;

exports[`HttpInterfaceVisitor > should ignore if return type if not a promise  1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const nest_http_interface_1 = require(\\"@r2don/nest-http-interface\\");
let UserService = class UserService {
    getUser() {
        throw new Error('not implemented');
    }
};
__decorate([
    (0, nest_http_interface_1.GetExchange)()
], UserService.prototype, \\"getUser\\", null);
UserService = __decorate([
    (0, nest_http_interface_1.HttpInterface)()
], UserService);
"
`;

exports[`HttpInterfaceVisitor > should ignore if return type is not a class 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const nest_http_interface_1 = require(\\"@r2don/nest-http-interface\\");
let TextService = class TextService {
    constructor() { }
    async getText() {
        return 'text';
    }
};
__decorate([
    (0, nest_http_interface_1.GetExchange)()
], TextService.prototype, \\"getText\\", null);
TextService = __decorate([
    (0, nest_http_interface_1.HttpInterface)()
], TextService);
"
`;

exports[`HttpInterfaceVisitor > should override plugin suffix option 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const r2don_http_module_1 = require(\\"@r2don/nest-http-interface\\");
const nest_http_interface_1 = require(\\"@r2don/nest-http-interface\\");
class ResponseClass {
}
let UserService = class UserService {
    async getUser() {
        throw new Error('not implemented');
    }
};
__decorate([
    (0, nest_http_interface_1.GraphQLExchange)(),
    r2don_http_module_1.ResponseBody(ResponseClass)
], UserService.prototype, \\"getUser\\", null);
UserService = __decorate([
    (0, nest_http_interface_1.HttpInterface)()
], UserService);
"
`;
